File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1; 
        int right = Integer.MIN_VALUE; 
        for(int p : piles){
            right = Math.max(right,p); 
        }

        int answer = 0; 
        while(left <= right){
            int k = left + (right - left) / 2; 

            long time = 0; 
            for(int n : piles){
                time += (n / k); 
                time += (n % k > 0 ? 1 : 0); 
            }

            if(time <= h){
                answer = k; 
                right = k - 1; 
            } else{
                left = k + 1; 
            }
        }

        return answer; 
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1 
        right = max(piles) 

        answer = 0

        while left <= right:
            k = left + (right - left) // 2 

            time = 0 
            for num in piles:
                time += (num // k) 
                time += 1 if num % k > 0 else 0

            if time <= h:
                answer = k  
                right = k - 1 
            else: 
                left = k + 1 

        return answer 
